sklearn/tests/test_multiclass.py

```diff
@@ -921,3 +922,108 @@ def test_ovo_consistent_binary_classification():
     ovo.fit(X, y)
 
     assert_array_equal(clf.predict(X), ovo.predict(X))
+
+
+# Test that metadata is routed correctly for multiclass-classifiers
+# =================================================================
+
+
+@pytest.mark.usefixtures("enable_slep006")
+def test_fitted_coefs_differ_when_sample_weight_differs():
+    """Assert if coefficients differ when estimators are fitted with different
+    `sample_weights`."""
+    sample_weight = np.ones_like(iris.target, dtype=np.float64)
+    multiclass_classifiers = [
+        OneVsRestClassifier,
+        OneVsOneClassifier,
+        OutputCodeClassifier,
+    ]
+    for classifier in multiclass_classifiers:
+        clf = classifier(
+            estimator=SGDClassifier(random_state=42).set_fit_request(sample_weight=True)
+        )
+        # fitting with sample_weight=sample_weight
+        clf.fit(iris.data, iris.target, sample_weight=sample_weight)
+        # TODO: should we raise an error if sample weight is requested but is not
+        # passed in the router's fit method when enable slep006 is True?
+        # clf.fit(iris.data, iris.target)
+
+        # fitting with sample_weight=sample_weight_new
+        clf_with_different_weight = clone(clf)
+        sample_weight_new = sample_weight.copy()
+        sample_weight_new[::2] /= 2
+        clf_with_different_weight.fit(
+            iris.data, iris.target, sample_weight=sample_weight_new
+        )
+
+        # test if coefficients are different
+        for est, est_with_different_weight in zip(
+            clf.estimators_, clf_with_different_weight.estimators_
+        ):
+            # we need "if hasattr" because OutputCodeClassifier
+            # might make use of class _ConstantPredictor
+            if hasattr(est, "coef_") and hasattr(est_with_different_weight, "coef_"):
+                assert not np.allclose(est.coef_, est_with_different_weight.coef_)
```

Comment:
While trying to relocate this test towards `sklearn/tests/test_metaestimators_metadata_routing.py`, this part here makes some trouble. There's some chance that `OutputCodeClassifier` will make use of `_ConstantPredictor` as one (or more) of its sub-estimators and will therefore not solely rely on the given sub-estimator. (It depends on how the classes are distributed, I believe.)
But `_ConstantPredictor` doesn't have a `registry` attribute implemented and thus every now an then  `test_setting_request_removes_error` in `sklearn/tests/test_metaestimators_metadata_routing.py` will raise at `assert registry`.

Possible patch: add `registry` attribute to `_ConstantPredictor`? IT's not used anywhere else, so it should not do any harm, but I feel that this is more of a dirty solution.

Reply:
I got you. And I think I didn't explain well in my above message. I will try again: I would delete the test in `test_multiclass.py`, if the central test n `est_metaestimators_metadata_routing.py` was working for `OutputCodeClassifier`, but it doesn't.

The reason is that `OutputCodeClassifier` sometimes internally uses a second sub-estimator (`_ConstantPredictor` ) that doesn't have a `registry` attribute, as required in the central tests. I was referring to the code snippet from `test_multiclass.py`, because this is how I had dealt with this issue for the test there.

We could make something similar for `est_metaestimators_metadata_routing.py`, or, easier but dirtier, implement a `registry` attribute into `_ConstantPredictor` (which is only used here and nowhere else).

But you need to tell me what is better or what other solution it could be.

Reply:
hmm, that's interesting. I think the solution here would be to make sure the input is constructed in a way that `_ConstantPredictor` is not used. So we can change the test and modify the data for this purpose.

Reply:
Or maybe this is not the reason for the fails, I'm starting to think. Still the test `test_setting_request_on_sub_estimator_removes_error` fails in 1/10 cases or so.

This is how it looks like:

```
               # sanity check that registry is not empty, or else the test passes
                # trivially
                print(f"\n\ninstance: {instance}")
                print(f"registry: {registry}")
>               assert registry
E               assert []
```

What I got printed can look like that (or similar), if this helps:

```
sklearn/tests/test_metaestimators_metadata_routing.py::test_setting_request_on_sub_estimator_removes_error[OutputCodeClassifier] 

instance: OutputCodeClassifier(estimator=ConsumingClassifier(registry=[ConsumingClassifier(registry=<Recursion on _Registry with id=139630032621456>),
                                                             ConsumingClassifier(registry=<Recursion on _Registry with id=139630032621456>)]))
registry: [ConsumingClassifier(('registry', [...])), ConsumingClassifier(('registry', [...]))]


instance: OutputCodeClassifier(estimator=ConsumingClassifier(registry=[]))
registry: []
```

Please, can you help here, @adrinjalali ?

Reply:
Okay: The test fails every time when the registry of the sub-estimators can't build at all, because the `OutputCodeClassifier` _solely_ uses `_ConstantPredictor`.

So, the idea was to change the data for testing, so `_ConstantPredictor` won't ever be used here.

I believe that this is not possible, because the `code_book_` used to make up binary classifications in `OutputCodeClassifier.fit()` is randomly generated. 

To see what I mean, run this code several times:

```
from sklearn.multiclass import OutputCodeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
X, y = make_classification(n_samples=100, n_features=4, n_informative=4, n_redundant=0, n_classes=3, random_state=0, shuffle=False)
clf = OutputCodeClassifier(estimator=RandomForestClassifier(random_state=0), code_size=1.5).fit(X, y)
print(f"classes_: {clf.classes_}\n")
print(f"code_book_: \n{clf.code_book_}\n")
print(f"estimators_: {clf.estimators_}\n")
```

Every time the randomly generated `code_book_` has a column of only 0 or 1 (column-wise binary classification), `_ConstantPredictor` is used as an estimator.

The only way to prevent this would be to set a `random_state` on `OutputCodeClassifier`, and this would heavily lower the quality of all the other tests, right?

Creating the test data with 3 instead of 2 classes (`y = rng.randint(0, 3, size=N)`  lowers the chances of all sub-estimators being `_ConstantPredictor` significantly, but it's still there.

I will push it, because I want to see if the CI likes it or not, but it's not a safe solution. Adding routing to `_ConstantPredictor` might be, I believe.

What you you think, @adrinjalali ?

Reply:
The CI didn't fail, but I believe it could have.

But now I've seen that I can set `init_args`  for individual meta-estimators. My comment before about this lowering overall test quality referred to putting a random state into the test, because I thought this was the only way.

Now I think setting `"init_args": {"random_state": 42}` for only for `OutputCodeClassifier` is a good idea. And also to keep testing with three classes, so the multiclass-classifiers actually do their jobs during these tests.